src/gazebo_mavlink_interface.cpp

```diff
@@ -644,17 +644,26 @@ void GazeboMavlinkInterface::setMavlinkSensorOrientation(const ignition::math::V
 
 void GazeboMavlinkInterface::ImuCallback(ImuPtr& imu_message)
 {
-  std::unique_lock<std::mutex> lock(last_imu_message_mutex_);
 
+  // Update imu message
   const int64_t diff = imu_message->seq() - last_imu_message_.seq();
   if (diff != 1 && imu_message->seq() != 0)
   {
     gzerr << "Skipped " << (diff - 1) << " IMU samples (presumably CPU usage is too high)\n";
   }
 
   last_imu_message_ = *imu_message;
-  lock.unlock();
-  last_imu_message_cond_.notify_one();
+
+  if (enable_lockstep_) {
+      std::unique_lock<std::mutex> lock(last_imu_message_mutex_);
+      lock.unlock();
+      last_imu_message_cond_.notify_one();
+  }
+  else {
+    SendSensorMessages();
+    SendGroundTruth();
+  }
+
```

Comment:
This lock needs to protect the write to `last_imu_message_`.
I couldn't remove L655 in the suggestion, but it should be removed.

```suggestion

  if (enable_lockstep_) {
      std::unique_lock<std::mutex> lock(last_imu_message_mutex_);
      last_imu_message_ = *imu_message;
      lock.unlock();
      last_imu_message_cond_.notify_one();
  }
  else {
    last_imu_message_ = *imu_message;
    SendSensorMessages();
    SendGroundTruth();
  }

```